bot/exts/info/tags.py

```diff
@@ -24,145 +29,229 @@
+        split_string = string.removeprefix(constants.Bot.prefix).split(" ", maxsplit=2)
+        if len(split_string) == 1:
+            return cls(None, split_string[0])
+        else:
+            return cls(split_string[0], split_string[1])
+
+
+class Tag:
+    """Provide an interface to a tag from resources with `file_content`."""
+
+    def __init__(self, content_path: Path):
+        post = frontmatter.loads(content_path.read_text("utf8"))
+        self.file_path = content_path
+        self.content = post.content
+        self.metadata = post.metadata
+        self._restricted_to: set[int] = set(self.metadata.get("restricted_to", ()))
+        self._cooldowns: dict[discord.TextChannel, float] = {}
+
+    @property
+    def embed(self) -> Embed:
+        """Create an embed for the tag."""
+        embed = Embed.from_dict(self.metadata.get("embed", {}))
+        embed.description = self.content
+        return embed
+
+    def accessible_by(self, member: discord.Member) -> bool:
+        """Check whether `member` can access the tag."""
+        return bool(
+            not self._restricted_to
+            or self._restricted_to & {role.id for role in member.roles}
+        )
+
+    def on_cooldown_in(self, channel: discord.TextChannel) -> bool:
+        """Check whether the tag is on cooldown in `channel`."""
+        return channel in self._cooldowns and self._cooldowns[channel] > time.time()
+
+    def set_cooldown_for(self, channel: discord.TextChannel) -> None:
+        """Set the tag to be on cooldown in `channel` for `constants.Cooldowns.tags` seconds."""
+        self._cooldowns[channel] = time.time() + constants.Cooldowns.tags
+
+
+def _fuzzy_search(search: str, target: str) -> float:
+    """A simple scoring algorithm based on how many letters are found / total, with order in mind."""
+    _search = REGEX_NON_ALPHABET.sub("", search.lower())
+    _targets = iter(REGEX_NON_ALPHABET.split(target.lower()))
+
+    current = 0
+    for _target in _targets:
+        index = 0
+        try:
+            while index < len(_target) and _search[current] == _target[index]:
+                current += 1
+                index += 1
+        except IndexError:
+            # Exit when _search runs out
+            break
+
+    return current / len(_search)
+
+
 class Tags(Cog):
-    """Save new tags and fetch existing tags."""
+    """Fetch tags by name or content."""
 
     def __init__(self, bot: Bot):
         self.bot = bot
-        self.tag_cooldowns = {}
-        self._cache = self.get_tags()
-
-    @staticmethod
-    def get_tags() -> dict:
-        """Get all tags."""
-        cache = {}
+        self.tags: dict[TagIdentifier, Tag] = {}
+        self.initialize_tags()
 
+    def initialize_tags(self) -> None:
+        """Load all tags from resources into `self.tags`."""
         base_path = Path("bot", "resources", "tags")
+
         for file in base_path.glob("**/*"):
             if file.is_file():
```

Comment:
Technically we can skip a level of indentation if we invert this

Reply:
The nesting is not too bad and there's nothing after it so I think it's fine as is